app/controllers/application_controller.rb

```diff
@@ -5,6 +5,159 @@ class ApplicationController < Sinatra::Base
+    if is_logged_in?
+      @tweet = Tweet.find_by_id(params[:id])  #find the tweet
+      if @tweet.user_id == session[:user_id] #check that the logged in user is the tweet owner
+       erb :'tweets/edit_tweet'   #then allow them to edit
+      else
+        redirect to '/tweets'  #then take back to user's tweets
+      end
+    else
+      redirect to '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    if params[:content] == '' #don't allow them update to a blank tweet
+      redirect to "/tweets/#{params[:id]}/edit"
+    else
+      @tweet = Tweet.find_by_id(params[:id]) #get the tweet
+      @tweet.content = params[:content]  #change the content to params
+      @tweet.save  #save the tweet to the db
+      redirect to "/tweets/#{@tweet.id}" #show the tweet
+    end
+  end
+
+  # DELETE
+  delete '/tweets/:id/delete' do
+    if is_logged_in?
+      @tweet = Tweet.find_by_id(params[:id])  #find that tweet
+      if @tweet.user_id == session[:user_id] #check that the person is the creator
+        @tweet.delete #destroy would work as well?
+        redirect to '/tweets'  #back to main user page
+      else
+        redirect to '/tweets' #if they are logged in but not their tweet, they can't delete
+      end
+    else
+      redirect to '/login'
+    end
+  end
+
+  # See individual user's tweets don't have to be logged in
+ get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+  ################ USER ###############
+  # User Sign up
+  get '/signup' do #Displays user sign up page unless person is already logged in
+    if is_logged_in?
+      redirect to '/tweets'
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do #processes the signup form submission
+    if params[:username] == "" || params[:email] == "" || params[:password] == "" # no empties allowed
+      redirect to '/signup' #failure page or message would be nice
+    else
+      @user = User.create(username: params[:username], email: params[:email], password: params[:password])
+      @user.save  #saves to DB
+      session[:user_id] = @user.id #log them in
+      redirect to '/tweets' #take them to their tweets
+    end
+  end
+
+  # USER LOG IN
+  get '/login' do #loads login page
+    if is_logged_in?
+      redirect to '/tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  post "/login" do   
+    @user = User.find_by(username: params[:username]) #find the user
+    if @user && @user.authenticate(params[:password]) #check password matches
+      session[:user_id] = @user.id   #log them in
+      redirect "/tweets" #show them some tweets
+    else
+      redirect "/login"
+    end
+  end
+
+  # USER LOGOUT
+  get '/logout' do
+    if is_logged_in?
+      session.clear   #logging out
+      redirect '/login'
+    else
+      redirect '/'
+    end
+  end
+
+  # Helpers
+  def is_logged_in?
+    !!session[:user_id]
+  end
+
+  def current_user
+    User.find(session[:user_id])
```

Comment:
Because in the lifespan of a request, you might call `current_user` multiple times, we'd prefer this method to only hit the database once and then 'cache' the user instance into an instance variable that could be returned on any subsequent call to `current_user`.

```ruby
def current_user
  @current_user ||= User.find(session[:id) if session[:id]
end
```

This pattern of conditionally setting an instance variable or returning it if it's already defined using the `||=` (or-equals) operator is called Memoization. ([read more](http://gavinmiller.io/2013/basics-of-ruby-memoization/) and [here](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/))